blivet/formats/disklabel.py

```diff
@@ -218,12 +218,71 @@ def parted_device(self):
             log.info("DiskLabel.parted_device returning None")
         return self._parted_device
 
+    @classmethod
+    def get_platform_label_types(cls):
+        label_types = ["msdos", "gpt"]
+        if arch.is_pmac():
+            label_types = ["mac"]
+        elif arch.is_efi() and not arch.is_aarch64():
+            label_types = ["gpt"]
+        elif arch.is_s390():
+            label_types = ["msdos"]  # since 'dasd' is only for DASD, it isn't listed here
+
+        return label_types
+
+    @classmethod
+    def set_default_label_type(cls, labeltype):
+        cls._default_label_type = labeltype
+        log.debug("default disklabel has been set to %s", labeltype)
+
+    def _label_type_size_check(self, label_type):
+        if self.parted_device is None:
+            return False
+
+        _label = parted.freshDisk(device=self.parted_device, ty=label_type)
+        return self.parted_device.length < _label.maxPartitionStartSector
+
+    def _get_best_label_type(self):
+        label_type = self.__class__._default_label_type
+        label_types = self.__class__.get_platform_label_types()[:]
+        if label_type in label_types:
+            label_types.remove(label_type)
+        if label_type:
+            label_types.insert(0, label_type)
+
+        if arch.is_s390():
+            if blockdev.s390.dasd_is_fba(self.device):
+                # the device is FBA DASD
+                return "msdos"
+            elif self.parted_device.type == parted.DEVICE_DASD:
+                # the device is DASD
+                return "dasd"
+
+        for lt in label_types:
+            if self._label_type_size_check(lt):
+                log.debug("selecting %s disklabel for %s based on size",
+                          label_type, os.path.basename(self.device))
+                label_type = lt
+                break
```

Comment:
Why not just ``return lt`` here?

Reply:
I just prefer to have one return statement per method, aside from very-early returns due to precondition checking.

Reply:
I see I returned early in the previous s390-specific block just above this one. It was just too much change to try to cram that extremely platform-specific logic into the single-return pattern.